test/integration/integration_test.cc

```diff
@@ -490,6 +524,58 @@ TEST_P(IntegrationTest, Pipeline) {
   connection.close();
 }
 
+// Checks to ensure that we reject the second request that is pipelined in the
+// same request
+TEST_P(IntegrationTest, PipelineWithTrailers) {
+  config_helper_.addConfigModifier(setEnableDownstreamTrailersHttp1());
+  config_helper_.addConfigModifier(setEnableUpstreamTrailersHttp1());
+  autonomous_upstream_ = true;
+  initialize();
+  std::string response;
+
+  std::string good_request("POST / HTTP/1.1\r\n"
+                           "Host: host\r\n"
+                           "Transfer-Encoding: chunked\r\n\r\n"
+                           "4\r\n"
+                           "body\r\n0\r\n"
+                           "trailer1:t2\r\n"
+                           "trailer2:t3\r\n"
+                           "\r\n");
+
+  std::string bad_request("POST / HTTP/1.1\r\n"
+                          "Host: host\r\n"
+                          "Transfer-Encoding: chunked\r\n\r\n"
+                          "4\r\n"
+                          "body\r\n0\r\n"
+                          "trailer1\r\n"
+                          "trailer2:t3\r\n"
+                          "\r\n");
+
+  Buffer::OwnedImpl buffer(absl::StrCat(good_request, good_request, bad_request));
+
+  RawConnectionDriver connection(
+      lookupPort("http"), buffer,
+      [&](Network::ClientConnection&, const Buffer::Instance& data) -> void {
+        response.append(data.toString());
+      },
+      version_);
+  // First response should be success.
+  size_t pos;
+  while ((pos = response.find("200")) == std::string::npos) {
+    connection.run(Event::Dispatcher::RunType::NonBlock);
+  }
+  EXPECT_THAT(response, HasSubstr("HTTP/1.1 200 OK\r\n"));
+  while (response.find("200", pos + 1) == std::string::npos) {
+    connection.run(Event::Dispatcher::RunType::NonBlock);
+  }
+  while (response.find("400") == std::string::npos) {
+    connection.run(Event::Dispatcher::RunType::NonBlock);
+  }
+
+  EXPECT_THAT(response, HasSubstr("HTTP/1.1 400 Bad Request\r\n"));
+  connection.close();
```

Comment:
Got a pipeline trailer integration test similar to the one above. However it is failing due to the autonomous connection not closing. I haven't had much time to figure out what it is, but if anyone knows it more than I do, please share 😄 . Otherwise will take a look tomorrow.

Reply:
At a guess (haven't looked in detail) I wonder if some part is not handling trailers as expected (maybe upstream codec?)  You could try turning on trace logging and seeing where things get stalled, or you can do some cerr logging

Reply:
Looking at the trace, it is getting a codec error and subsequently is resetting the stream which is what we want.

Reply:
Nevermind, I'll just need to add support to autonomous upstream to support disconnects.

Reply:
Will need this PR in https://github.com/envoyproxy/envoy/pull/9148/files